bionic/code_analyzer.py

```diff
@@ -0,0 +1,203 @@
+"""
+This module implements the logic to encode your code objects into a
+byte array that unqiuely identifies the objects. Any changes to the
+code objects would result in a different byte array encoding. It looks
+for any changes to the function bytecode and function constants for
+now. But in the future, we will add more functionality like checking
+function references and any changes to class objects.
+
+It is used by Bionic to detect any changes to entity functions. When
+the entity function changes, the value of the Bionic entity can change,
+which requires Bionic to invalidate the cache and compute the entity
+again.
+
+This logic is only used when versioning mode is set to "auto" or
+"assisted". Since the logic doesn't cover every type of code change,
+like references or classes, this logic is still experimental. When it
+reaches a state where it detects all kinds of changes, we will make
+this the default behavior.
+"""
+
+import inspect
+import json
+import warnings
+
+from .utils.misc import oneline
+
+
+# We use object names for circular references when analyzing code. For
```

Comment:
That doesn't work since object names aren't detecting circular refs. But I get your point that this by itself isn't giving enough info. I'll change it to, "We analyze objects names instead of the objects when they are circularly referenced."